### DISCUSSION TITLE
Trouble tuning Bolt 1.1 + F450

### DISCUSSION THREAD

**Discussion by johnnwallace** [bitcraze/discussions#150] [Q&A]
I am working on a drone powered by the [Bolt 1.1](https://store.bitcraze.io/products/crazyflie-bolt-1-1) on an [F450](https://www.amazon.com/YoungRC-4-Axis-Airframe-Quadcopter-Landing/dp/B0776WLHX7/ref=sr_1_5?crid=1M22727HYG4TB&keywords=dji+f450&qid=1652326310&sprefix=dji+f450%2Caps%2C58&sr=8-5) frame. Has anyone had experience using the Bolt on such a frame? It would be awesome if I had some rate PID gains to jump off with. We are using 1400 KV motors on a 4S. I don't expect anyone to have the exact same configuration, but I figured using the same gains as another F450 drone would be a good starting point.

If anyone has any other advice on tuning, it would be greatly appreciated. Our initial tries getting the drone off the ground resulted in many examples of the drone quickly flipping on its pitch axis, like [this](https://youtu.be/cnkYNpp2CM4) one. Following the advice on slide 18 of [this](https://www.bitcraze.io/about/events/documents/bam2021/BQ_Bolt_workshop_bam2021.pdf) guide, I halved the roll gains from the CF 2.1. That resulted in [this](https://youtube.com/shorts/H3CUlsgannI?feature=share) test, where we get a very unstable hover. I haven't been able to reproduce similar behavior when using the same gains, instead getting the same flips over the pitch axis as before.

Any advice would be awesome!

edit 1: I have seen [this](https://forum.bitcraze.io/viewtopic.php?t=4474) post, but no gains

edit 2: We actually got the drone to more consistently get up in the air by dramatically decreasing the thrustBase parameter (down to 15000), which is well below the thrust needed to hover ([video](https://youtube.com/shorts/w9TKViZ97Fs)). We tried tuning some more, but we found that gains too low would not correct for a pitch over that occurs while trying to take off (still looking into why this happens), while high enough gains to take off would lead to a very jittery flight.

A potential option would be to decrease gains once takeoff is detected, but ideally, we would try to figure out why the early pitch-over occurs. It doesn't seem to be a motor issue, since we tried swapping the motors and the same pitch-over occurs, and we already calibrated the ESCs. It's possible that some of the ESCs have trouble ramping up quickly for take-off also, I guess.

**Comment by knmcguire** [comment]
Hi!

I have a couple of comments:

1. I'm not sure that using the gains from other platforms will help as the crazyflie controllers are not really standardized. Could you perhaps share the gains here?
2. I've contacted the person from the forum post to share the gains as well so hopefully he will answer soon.
3. Also, I'm pretty sure the slides made a typo there, as you should take halve of the roll AND pitch gains. I think you should always tune those symmetrically. 
4. The latest development branch of the cfclient has a tuning tab now!  So if you install the [latest version from source](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/installation/install/), you should be able to try it out.

**Reply by knmcguire** [reply]
An additional comment,  the platform on the slides is only twice as big as a crazyflie (about 20 cm diameter). You should probably tune de gains down even more than half for a big platform like the f450

**Reply by johnnwallace** [reply]
Thanks for the advice! We actually got the drone to more consistently get up in the air by dramatically decreasing the thrustBase parameter (down to 15000), which is well below the thrust needed to hover ([video](https://youtube.com/shorts/w9TKViZ97Fs)). We tried tuning some more, but we found that gains too low would not correct for a pitch over that occurs while trying to take off (still looking into why this happens), while high enough gains to take off would lead to a very jittery flight.

A potential option would be to decrease gains once takeoff is detected, but ideally, we would try to figure out why the early pitch-over occurs. It doesn't seem to be a motor issue, since we tried swapping the motors and the same pitch-over occurs, and we already calibrated the ESCs. It's possible that some of the ESCs have trouble ramping up quickly for take-off also, I guess.

**Comment by knmcguire** [comment]
I managed to get an answer of the user of that forum post that you referred to! These were the values he ended up with down here. 

Mind that he said that he was only able to tune the innerloop properly (pitch roll yaw), but had still problems with the outer loop (Velocities positions) including the altitude hold. So just be aware of that!

Here are the innerloop gains:

```
#define PID_ROLL_RATE_KP  100.0
#define PID_ROLL_RATE_KI  100.0
#define PID_ROLL_RATE_KD  1.0
#define PID_ROLL_RATE_INTEGRATION_LIMIT    33.3

#define PID_PITCH_RATE_KP  100.0
#define PID_PITCH_RATE_KI  100.0
#define PID_PITCH_RATE_KD  1.0
#define PID_PITCH_RATE_INTEGRATION_LIMIT   33.3

#define PID_YAW_RATE_KP  60.0
#define PID_YAW_RATE_KI  8.3
#define PID_YAW_RATE_KD  0.0
#define PID_YAW_RATE_INTEGRATION_LIMIT     166.7

#define PID_ROLL_KP  3.0
#define PID_ROLL_KI  1.5
#define PID_ROLL_KD  0.0
#define PID_ROLL_INTEGRATION_LIMIT    20.0

#define PID_PITCH_KP  3.0
#define PID_PITCH_KI  1.5
#define PID_PITCH_KD  0.0
#define PID_PITCH_INTEGRATION_LIMIT   20.0

#define PID_YAW_KP  6.0
#define PID_YAW_KI  1.0
#define PID_YAW_KD  0.35
#define PID_YAW_INTEGRATION_LIMIT     360.0


#define DEFAULT_PID_INTEGRATION_LIMIT 5000.0
#define DEFAULT_PID_OUTPUT_LIMIT      0.0
```
